plugins: unique command line options, small fixes and plugin documentation

SimonVrouwe opened the pull request:
Some issues I came across looking at #5294 (that discussion is ongoing, this PR adds some context)

Converts `plugin->cmd` to a full path immediately (couldn't find a reason not to), ensures only one instance can run.
Plugin command line options should have unique names (and no aliases).
Update plugin documentation.

**edit** ~Last 2~ [8be7693](https://github.com/ElementsProject/lightning/pull/5343/commits/8be769345a6b3b869022bcb7ae3cd107ffb162d5) and  [cf2dbbb](https://github.com/ElementsProject/lightning/pull/5343/commits/cf2dbbbe6a71b6a383f5126778e74bd915443641) ~commits is an idea for plugins that do configs internally, not fully worked out.~

**edit** Add "dynamic" field to `plugin list` result, some builtin plugins are dynamic !

vincenzopalazzo commented on lightningd/plugin.c:
```diff
@@ -250,11 +250,17 @@ struct plugin *plugin_register(struct plugins *plugins, const char* path TAKES,
 	u32 chksum;
 
+	abspath = path_canon(plugins, path);
+	if (!abspath) {
+		tal_free(abspath);
+		return NULL;
+	}
```
```suggestion
	if (!abspath)
		return tal_free(abspath);
```

vincenzopalazzo reviewed (commented):
Concept ACK.

I need to return on it!

SimonVrouwe commented:
About halve of the builtin plugins are "dynamic", is this deliberate?
Calling `plugin stop` on them will shutdown `lightningd`.

cdecker commented on lightningd/plugin.c:
```diff
@@ -267,13 +273,14 @@ struct plugin *plugin_register(struct plugins *plugins, const char* path TAKES,
 	p = tal(plugins, struct plugin);
 	p->plugins = plugins;
-	p->cmd = tal_strdup(p, path);
+	p->cmd = tal_steal(p, abspath);
```
Looks like the `path` is no longer freed (`tal_strdup` would have stolen it instead of copying when `take(path)` was used). This can leak now if it is taken.

SimonVrouwe replied:
Isn't `path` already taken earlier in
```
abspath = path_canon(plugins, path);
```

Or maybe not when it failed?
https://github.com/ElementsProject/lightning/blob/9152b8c4240957f185ea026c135eccf335b62069/ccan/ccan/tal/path/path.h#L26-L34

SimonVrouwe replied:
Yes, it will take it (below the commented-out code):
https://github.com/ElementsProject/lightning/blob/bad943da55b367158f5e080baea90754aa50c9a7/ccan/ccan/tal/path/path.c#L215-L216

cdecker commented on lightningd/plugin.c:
```diff
@@ -930,6 +930,9 @@ static const char *plugin_opt_add(struct plugin *plugin, const char *buffer,
 	popt->name = tal_fmt(popt, "--%.*s", nametok->end - nametok->start,
 			     buffer + nametok->start);
+	/* an "|" alias could circumvent the unique-option-name check */
+	if (!deprecated_apis && tal_strreg(tmpctx, popt->name, "[|]"))
```
I wonder why `strchr` isn't a better (more performant) choice here?

SimonVrouwe replied:
Yes, I wasn't sure if we need to sanitize more, e.g. like "[^a-zA-Z0-9_-]" regex? Otherwise I'm happy to change it.

Currently setting an option like `~!@#$%^&*()+[{]};:=hello` in the config file just works.
I couldn't find any spec for valid `--longopts` chars.

rustyrussell replied:
strchr here is fine. We assume plugins are not malicious :)

cdecker commented on lightningd/plugin.h:
```diff
@@ -88,6 +88,9 @@ struct plugin {
 	 * and that other plugins may subscribe to. */
 	const char **notification_topics;
+
+	/* Plugin addendum (e.g. configs) it wants shown by `listconfigs` */
+	const char *addendum;
```
Going to be quite horrible for schema-based RPC bindings, but I like the idea.

rustyrussell replied:
I would resist this strongly until proven necessary.

It's not consistent to have things which are not configs in listconfigs.  If it's in listconfigs (and, yes, adding plugin randomness to listconfigs was my mistake), it should be valid as cmdline/config file.

SimonVrouwe replied:
> I would resist this strongly until proven necessary.

Agreed,  will drop it.
Maybe something for another PR and then add it to `plugin list` instead of `listconfigs`.

cdecker commented on doc/schemas/listconfigs.schema.json:
```diff
@@ -33,6 +33,16 @@
             "description": "Specific options set for this plugin",
             "properties": {}
+          },
+          "internal_config": {
+            "type": "array",
+            "items": {
+              "type": "object",
+              "additionalProperties": true,
+              "required": [],
+              "description": "Specific internal option of this plugin",
+              "properties": {}
+            }
```
Wouldn't it be better to just allow this to be a dict? Then plugins can just have key-value pairs instead of having to nest them into an array, which I think is the most common thing plugins will do. Alternatively we could allow `additionalProperties` on wrapping object, though that'd mean losing some checks.

SimonVrouwe replied:
Dropped this part, probably for another PR. Yes dict would be nice.

cdecker reviewed (approved):
Great cleanup of some longstanding issues :+1:

rustyrussell reviewed (commented):
Sorry, these comments were sitting pending in GH for a while :(

SimonVrouwe commented:
Applied the feedback and some minor fixes, squashed and rebased.
The `addendum` stuff has been dropped (for now).
I added some relative links and a footnote to PLUGINS.md, hopefully these work in readthedocs.

SimonVrouwe commented:
The most serious test fail seems a [memleak](https://github.com/ElementsProject/lightning/runs/7183689223?check_suite_focus=true) in `test_openchannel_hook_chaining`:
```
E            - Node /tmp/ltests-y28mxbp5/test_openchannel_hook_chaining_1/lightning-2/ has memory leaks: [
E               {
E                   "backtrace": [
E                       "ccan/ccan/tal/tal.c:442 (tal_alloc_)",
E                       "ccan/ccan/io/io.c:91 (io_new_conn_)",
E                       "lightningd/subd.c:774 (new_subd)",
E                       "lightningd/subd.c:828 (new_channel_subd_)",
E                       "lightningd/opening_control.c:872 (peer_start_openingd)",
E                       "lightningd/peer_control.c:1311 (peer_active)",
E                       "lightningd/connect_control.c:458 (connectd_msg)",
E                       "lightningd/subd.c:557 (sd_msg_read)",
E                       "lightningd/subd.c:357 (read_fds)",
E                       "ccan/ccan/io/io.c:59 (next_plan)",
E                       "ccan/ccan/io/io.c:407 (do_plan)",
E                       "ccan/ccan/io/io.c:417 (io_ready)",
E                       "ccan/ccan/io/poll.c:453 (io_loop)",
E                       "lightningd/io_loop_with_timers.c:22 (io_loop_with_timers)",
E                       "lightningd/lightningd.c:1182 (main)",
E                       "../csu/libc-start.c:308 (__libc_start_main)"
E                   ],
E                   "label": "ccan/ccan/io/io.c:91:struct io_conn",
E                   "parents": [
E                       "lightningd/lightningd.c:107:struct lightningd"
E                   ],
E                   "value": "0x55d72cfc3058"
```

```
2022-07-04T17:12:53.278Z DEBUG   lightningd: openchannel hook rejects and says 'reject on principle'
2022-07-04T17:12:53.278Z DEBUG   0266e4598d1d3c415f572a8488830b60f7e744ed9235eb0b1ba93283b315c03518-openingd-chan#2: peer_out WIRE_ERROR
2022-07-04T17:12:53.279Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-openingd-chan#2: peer_in WIRE_ERROR
2022-07-04T17:12:53.279Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-openingd-chan#2: aborted opening negotiation: They sent error channel ae288ec6893205a5f369f192a924252d31c7cec52c2f05da658d1ace512ae795: You gave bad parameters: reject on principle
2022-07-04T17:12:53.279Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-openingd-chan#2: billboard perm: They sent error channel ae288ec6893205a5f369f192a924252d31c7cec52c2f05da658d1ace512ae795: You gave bad parameters: reject on principle
2022-07-04T17:12:53.280Z DEBUG   plugin-spenderp: mfc 33, dest 0: failed! fundchannel_start 022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59: {\"code\":-1,\"message\":\"They sent error channel ae288ec6893205a5f369f192a924252d31c7cec52c2f05da658d1ace512ae795: You gave bad parameters: reject on principle\"}.
2022-07-04T17:12:53.280Z DEBUG   plugin-spenderp: mfc 33: parallel channel starts done.
2022-07-04T17:12:53.280Z DEBUG   plugin-spenderp: mfc 33: trying redo despite 'fundchannel_start' failure (They sent error channel ae288ec6893205a5f369f192a924252d31c7cec52c2f05da658d1ace512ae795: You gave bad parameters: reject on principle); will cleanup for now.
2022-07-04T17:12:53.280Z DEBUG   plugin-spenderp: mfc 33: cleanup!
2022-07-04T17:12:53.280Z DEBUG   plugin-spenderp: mfc 33: unreserveinputs task.
2022-07-04T17:12:53.282Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-connectd: destroy_subd: 1 subds, to_peer conn 0x55782e6644b8, read_to_die = 0
2022-07-04T17:12:53.282Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-connectd: disconnect
2022-07-04T17:12:53.282Z DEBUG   022d223620a359a47ff7f7ac447c85c46c923da53389221a0054c11c1e3ca31d59-connectd: peer_conn_closed
2022-07-04T17:12:53.282Z DEBUG   0266e4598d1d3c415f572a8488830b60f7e744ed9235eb0b1ba93283b315c03518-connectd: destroy_subd: 1 subds, to_peer conn (nil), read_to_die = 0
```

Note the `nil` logged by L2 in the last line, it seems `io_conn` was already taken/moved ?

Will try to reproduce, but I don't think the problem lies in this PR but maybe it triggered a race.

niftynei commented:
> The most serious test fail seems a [memleak](https://github.com/ElementsProject/lightning/runs/7183689223?check_suite_focus=true) in test_openchannel_hook_chaining:

This is failing elsewhere, is a general regression on master.

niftynei commented:
ACK https://github.com/ElementsProject/lightning/pull/5343/commits/d267f270fffc1e62ac1d28b86276fd9905636fda

Test failures are unrelated, i've saved the logs from the `fetchinvoice` error and will add a separate issue to track the flakes.
